src/wasm/wasm-validator.cpp

```diff
@@ -2199,6 +2152,19 @@ void FunctionValidator::visitTupleExtract(TupleExtract* curr) {
   }
 }
 
+void FunctionValidator::visitCallRef(CallRef* curr) {
```

Comment:
It would be good to check that the tail-call feature is enabled if this is a `return_call_ref`, like `visitCall` and `visitCallIndirect` do. It may be worth abstracting that out, too.

Reply:
Good idea, refactoring.